File: backend-services/product-and-inventory/src/common/mappers/product.mapper.ts
import { ProductDocument } from '../../modules/products/schemas/product.schema';
import { ProductResponseDto } from '../../modules/products/dto/responses/product-response.dto';

export function mapProductToResponseDto(
  product: ProductDocument,
): ProductResponseDto {
  return {
    id: product._id.toString(),
    vendor_id: product.vendor_id,
    category_id: product.category_id?.toString() || '',
    name: product.name,
    price: product.price,
    stock: product.stock,
    image: product.image,
  };
}

File: backend-services/inventory_go/product/mappers/product_mapper.go
package mappers

import (
	"go.mongodb.org/mongo-driver/v2/bson"
	"inventory_go/infrastructure/models"
	"inventory_go/product"
)

func ProductDomainToDocument(product *product.Product) (*models.ProductDocument, error) {
	var id bson.ObjectID
	var err error

	if product.Id() != "" {
		id, err = bson.ObjectIDFromHex(product.Id())
		if err != nil {
			return nil, err
		}
	}

	var categoryId bson.ObjectID
	if product.CategoryId() != "" {
		categoryId, err = bson.ObjectIDFromHex(product.CategoryId())
		if err != nil {
			return nil, err
		}
	}

	return &models.ProductDocument{
		Id:          id,
		CategoryId:  categoryId,
		Name:        product.Name(),
		Price:       product.Price(),
		Stock:       product.Stock(),
		VendorId:    product.VendorId(),
		Image:       product.Image(),
		Description: product.Description(),
	}, nil
}

func ProductDocumentToDomain(document *models.ProductDocument) *product.Product {
	var id string
	if !bson.ObjectID.IsZero(document.Id) {
		id = document.Id.Hex()
	}

	var categoryId string
	if !bson.ObjectID.IsZero(document.CategoryId) {
		categoryId = document.CategoryId.Hex()
	}
	return product.NewProduct(id, document.Name, document.Price, document.Stock, document.VendorId, document.Image, categoryId, document.Description)

}
